cluster-autoscaler/clusterstate/clusterstate.go

```diff
@@ -605,14 +620,17 @@ func (csr *ClusterStateRegistry) updateReadinessStats(currentTime time.Time) {
 	perNodeGroup := make(map[string]Readiness)
 	total := Readiness{Time: currentTime}
 
-	update := func(current Readiness, node *apiv1.Node, nr kube_util.NodeReadiness) Readiness {
+	update := func(current Readiness, node *apiv1.Node, nr kube_util.NodeReadiness, nodeGroup cloudprovider.NodeGroup) Readiness {
 		current.Registered = append(current.Registered, node.Name)
 		if _, isDeleted := csr.deletedNodes[node.Name]; isDeleted {
 			current.Deleted = append(current.Deleted, node.Name)
 		} else if nr.Ready {
 			current.Ready = append(current.Ready, node.Name)
 		} else if node.CreationTimestamp.Time.Add(MaxNodeStartupTime).After(currentTime) {
 			current.NotStarted = append(current.NotStarted, node.Name)
+		} else if cloudprovider.IsNodeGroupHibernateEnabled(nodeGroup) && taints.HasShutdownTaint(node) || taints.HasUnreachableTaint(node) {
```

Comment:
Relying on generic taints here seems quite fragile. What if the node is unreachable for other reasons than hibernation? Misconfigured firewall rules, perhaps? Since you introduced `InstanceHibernating` and `InstanceHibernated` states for instances, why not rely on that instead?